src/main/java/frc/robot/subsystems/targetvision/TargetVision.kt

```diff
@@ -1,61 +1,203 @@
+        score += (TY_WEIGHT)*( 1 - (deltaY.pow(TY_DROPOFF_RATE) / 41.0.pow(TY_DROPOFF_RATE)) ) //20.5 = vertical fov of limelight
+        score += (TA_WEIGHT)*( 1 - (deltaA.pow(TA_DROPOFF_RATE) / TA_MAX.pow(TA_DROPOFF_RATE)) )
+
+        return score
     }
 
-    private fun addMeasurement(measurement: Measurement) {
-        measurements.add(measurement)
 
-        if (measurements.size > SAMPLE_NUM) {
-            measurements.removeFirst()
+    fun addMeasurement(measurement: Measurement){
+
+        var grouped = false
+
+        // group targets
+        for (pose in robotPoses) {
+            // TODO: Find good threshold values for grouping
+            if(groupConfidence(pose.last().pose, measurement.pose) > GROUPING_CONFIDENCE_THRESHOLD) {
+                pose.add(measurement)
+                if ((measurement.timestamp - pose.first().timestamp > MAX_SAMPLE_TIME_MS)) { // TODO: find good value
+                    // cleanup old targets
+                    pose.removeFirst()
+                }
+                grouped = true
+                break
+            } else if((measurement.timestamp - pose.last().timestamp > MAX_SAMPLE_TIME_MS)){
+                robotPoses.remove(pose)
+            }
+        }
+
+        if(!grouped){
+            robotPoses.add(mutableListOf(measurement))
         }
+
+
+
+
+
+   }
+
+
+    //give a measurement a rating based on how likely it is to be our current primary target
+    private fun rateTrackingPreference(measurement: List<List<Measurement>>): Int{
+
+        fun getVelocity(samples: Pair<List<Measurement>, Int>): Double{
+            val tVector: List<Double> = samples.first.map{(timestamp, _) -> timestamp}
+            val txVector: List<Double> = samples.first.map{(_, target) -> target.tx}
+
+            // regressing tx
+            return tVector.zip(txVector).sumOf { pair -> pair.first * pair.second } /
+                    tVector.zip(tVector).sumOf { pair -> pair.first * pair.second }
+
+        }
+
+        var measurementIndexed = measurement.map { list -> Pair(list, measurement.indexOf(list))}
+
+        //filter by tx
+        measurementIndexed = measurementIndexed.sortedBy { pair -> abs(pair.first.last().pose.tx)}
+        val lowestTx = abs(measurementIndexed[0].first.last().pose.tx)
+        measurementIndexed = measurementIndexed.filter{ pair -> abs(pair.first.last().pose.tx) in lowestTx..lowestTx+TX_THRESHOLD}
+
+        //filter by distance
+        measurementIndexed = measurementIndexed.sortedBy {pair -> getDistance(pair.first.last().pose)}
+        val lowestDistance = getDistance(measurementIndexed[0].first.last().pose)
+        measurementIndexed = measurementIndexed.filter{ pair -> getDistance(pair.first.last().pose) in lowestDistance..lowestDistance + DISTANCE_THRESHOLD}
+
+
+        return measurementIndexed.minByOrNull { getVelocity(it) }!!.second
```

Comment:
This non-null assertion is unsound. If no targets (measurements? samples? smaples? poses? I have no fucking clue what any of the names in this file mean.) meet the `tx` and distance thresholds the code will crash.

Reply:
there will always be a target that meets the tx / distance threshold because the threshold min is set to the lowest tx of any targets

Reply:
There could be zero targets.